### PR TITLE
Bug 1194091 - [browser] Firefox Account sync bookmark

### PR DESCRIPTION
**Pull request by danhuang1202**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ea37e50c14b3450cf19841d0282a6e0087474d0b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d4e652d641934ea02dcaec4fb9607fddfe0e8702

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d17e5752a885660604e640a6c092ea95a4aef39c

File: tv_apps/browser/js/bookmarkStore.js [hunk 1]
```diff
@@ -1,34 +1,109 @@
...
+    isSynced: false,
+
     cache: [],
 
+    currentFolder : null,
+
+    reset: function(cb) {
+      this.cache = [];
+      this.currentFolder = null;
+      SyncManagerBridge.getInfo().then(message => {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please use Promise instead. For example ```return SyncManagerBridge.getInfo().then(message => {```.

File: tv_apps/browser/js/awesomescreen.js [hunk 2]
```diff
@@ -655,7 +658,13 @@ var Awesomescreen = {
    * Select the Bookmarks tab.
    */
   selectBookmarksTab: function awesomescreen_selectBookmarksTab() {
-    BookmarkStore.updateCache(this.bookmarkList.open.bind(this.bookmarkList));
+    BookmarkStore.reset(
```
Comment (begeeben) [inline tv_apps/browser/js/awesomescreen.js]: Once the ```reset``` function is changed to return a ```Promise```, please also modify the code here.

File: tv_apps/browser/js/bookmarkStore.js [hunk 3]
```diff
@@ -1,34 +1,109 @@
...
+        this.isSynced = (message.state === 'enabled') ? true : false;
+        cb();
+      });
+    },
+
     getByRange: function(start, num, folderId, cb) {
-      var i = start,
+      var fn = function(){
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please name the function by what it does and change the style to ```function functionName()```.

File: tv_apps/browser/js/bookmarkStore.js [hunk 4]
```diff
@@ -1,34 +1,109 @@
...
+        var i = start,
           length = (start + num) > this.cache.length ?
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Sorry that I missed this part last time. Please separate ```length``` and ```result``` with another 2 ```var``` definitions. For better readability.

File: tv_apps/browser/js/bookmarkStore.js [hunk 5]
```diff
@@ -1,34 +1,109 @@
...
               (this.cache.length - start) : (start + num),
           result = [];
-      for(; i < length; i++) {
-        result.push(this.cache[i]);
+        for(; i < length; i++) {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please define the ```var i``` here in the ```for``` loop for better readability.

File: tv_apps/browser/js/bookmarkStore.js [hunk 6]
```diff
@@ -1,34 +1,109 @@
...
+
+    reset: function(cb) {
+      this.cache = [];
+      this.currentFolder = null;
+      SyncManagerBridge.getInfo().then(message => {
+        this.isSynced = (message.state === 'enabled') ? true : false;
+        cb();
+      });
+    },
+
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please add a ```// TODO:``` for the live time bookmark data loading task.

File: tv_apps/browser/js/bookmarkStore.js [hunk 7]
```diff
@@ -1,34 +1,109 @@
...
 
     updateCache: function(cb){
-      BrowserDB.getBookmarks((function(bookmarks){
-        this.cache = bookmarks;
-        cb();
-      }).bind(this));
+      this.cache = [];
+
+      if(this.isSynced) {
+        if(!this.currentFolder) {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Ideally we should use ```Promise.all```. Please add a comment to indicate this.

File: tv_apps/browser/js/bookmarkStore.js [hunk 8]
```diff
@@ -1,34 +1,109 @@
...
+                b.readOnly = true;
+                this.cache.push(b);
+              }
+            });
+            cb();
+          });
+        }
+      } else {
+        // Get bookmark data from origin indexdDB
+        BrowserDB.getBookmarks(bookmarks => {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkStore.js]: Please move this block to the top of the function with 
```
if (!this.isSynced) {
  ...
  return;
}
```

File: tv_apps/browser/js/awesomescreen.js [hunk 9]
```diff
@@ -1737,7 +1754,11 @@ var Awesomescreen = {
       if(this.bookmarkList.isDisplay()) {
         BrowserDB.removeBookmark(
           listUrl,
-          this.bookmarkList.removeFocusItem.bind(this.bookmarkList)
+          function(){
+            BookmarkStore
+              .updateCache(
```
Comment (begeeben) [inline tv_apps/browser/js/awesomescreen.js]: Please concatenate this line with the line above.

File: tv_apps/browser/js/awesomescreen.js [hunk 10]
```diff
@@ -1777,7 +1798,13 @@ var Awesomescreen = {
       BrowserDB.updateBookmark(
         listUrl,
         bmTitle,
-        this.bookmarkList.updateFocusItemTitle.bind(this.bookmarkList, bmTitle)
+        function(){
+          BookmarkStore
+            .updateCache(
+              this
+                .bookmarkList
+                  .updateFocusItemTitle.bind(this.bookmarkList, bmTitle));
```
Comment (begeeben) [inline tv_apps/browser/js/awesomescreen.js]:
```

          BookmarkStore.updateCache(
            this.bookmarkList
                .updateFocusItemTitle.bind(this.bookmarkList, bmTitle));
```

File: tv_apps/browser/js/awesomescreen.js [hunk 11]
```diff
@@ -155,18 +156,16 @@ var Awesomescreen = {
...
     }).bind(this));
 
-    this.bookmark.addEventListener('itemUpdated', (function(e){
-      BookmarkStore.updateCache();
-      this.dialogHidden();
-    }).bind(this));
-
-    this.bookmark.addEventListener('itemRemoved', (function(e){
-      BookmarkStore.updateCache();
+    this.bookmark.addEventListener('hideDialog', (function(e){
```
Comment (begeeben) [inline tv_apps/browser/js/awesomescreen.js]: After I look deeper in the ```BookmarkList```, I found that replacing ```itemUpdated``` and ```itemRemoved``` with ```hideDialog``` is some how not logical. When a ```BookmarkList``` is being used by an outside module, the ```BookmarkList``` has no idea that the outside module has a dialog that needs to be hidden. Please stay with the origin design.

File: tv_apps/browser/js/bookmarkList.js [hunk 12]
```diff
@@ -472,8 +458,7 @@
           this.focusItem(this.listItemMap[this.focusIndex]);
         }
 
-        var event = new Event('itemRemoved');
-        this.el.dispatchEvent(event);
+        this.dispatchHideDialog();
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: As state above, please keep the original design.

File: tv_apps/browser/js/bookmarkList.js [hunk 13]
```diff
@@ -834,6 +854,11 @@
         }});
       this.el.dispatchEvent(event);
     },
+
+    dispatchHideDialog: function() {
+      var event = new Event('hideDialog');
+      this.el.dispatchEvent(event);
+    }
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: As explained above, please remove this method.

File: tv_apps/browser/js/bookmarkList.js [hunk 14]
```diff
@@ -610,13 +604,24 @@
 
       if(targetEl) {
         targetEl.querySelector('.title').textContent = title;
-        var event = new Event('itemUpdated');
-        this.el.dispatchEvent(event);
+        this.dispatchHideDialog();
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: As explained above, please keep the original design.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e29037e200d688de8cd0eaccbd29abeb29dfffbf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=75cc4d96c05bf865c5f2b78a75edb8b9f437c58e

**Pull request merged**